### DISCUSSION TITLE
Open API Pages: Totally Pythonic, OpenAPI Based customizable documentation pages for SwaggerUI, ReDoc, RapiDoc, Elements, Scalar.

### DISCUSSION THREAD

**Discussion by hasansezertasan** [fastapi/fastapi#11158] [Show and tell]
# [OpenAPIPages](https://github.com/hasansezertasan/openapipages) 📄

Hello everyone 👋.

I would like to talk about a package I built this week.

Totally Pythonic, OpenAPI Based customizable documentation pages for SwaggerUI, ReDoc, RapiDoc, Elements, Scalar.

## Features

> Gimme an OpenAPI Spec, leave the rest to me...

- Framework agnostic.
- Zero dependencies, just Python standard library.
- Fully typed.
- Highly extensible.

[OpenAPIPages](https://github.com/hasansezertasan/openapipages) follows the same structure as [FastAPI to create Documentation Pages](https://github.com/tiangolo/fastapi/blob/be876902554a0bd886167de144f0d593ed2e6ad7/fastapi/openapi/docs.py).

## Usage

Install with `pip install openapipages` command and you are ready to go...

```python
from fastapi import FastAPI
from fastapi.responses import HTMLResponse
from openapipages import Elements, RapiDoc, ReDoc, Scalar, SwaggerUI

app = FastAPI()


@app.get("/")
def root() -> dict[str, str]:
    return {"Hello": "World"}


@app.get("/swaggerui", response_class=HTMLResponse, include_in_schema=False)
def get_swaggerui() -> str:
    return SwaggerUI(title="Swagger UI").render()


@app.get("/redoc", response_class=HTMLResponse, include_in_schema=False)
def get_redoc() -> str:
    return ReDoc(title="ReDoc").render()


@app.get("/scalar", response_class=HTMLResponse, include_in_schema=False)
def get_scalar() -> str:
    return Scalar(title="Scalar").render()


@app.get("/elements", response_class=HTMLResponse, include_in_schema=False)
def get_elements() -> str:
    return Elements(title="Elements").render()


@app.get("/rapidoc", response_class=HTMLResponse, include_in_schema=False)
def get_rapidoc() -> str:
    return RapiDoc(title="RapiDoc").render()

```

In this small application, we have 5 different Documentation UIs.

Swagger UI:

![image](https://github.com/tiangolo/fastapi/assets/13135006/260bdb7f-d540-4b9e-b857-8e145a117569)

ReDoc UI:

![image](https://github.com/tiangolo/fastapi/assets/13135006/8397af92-603d-413f-8694-efa5dac44312)

RapiDoc UI:

![image](https://github.com/tiangolo/fastapi/assets/13135006/758572e5-5dc9-4601-8874-e50094272da0)

Elements UI:

![image](https://github.com/tiangolo/fastapi/assets/13135006/c816c1af-d375-4556-91e3-3135bcd01bb4)

And finally...

Scalar UI:

![image](https://github.com/tiangolo/fastapi/assets/13135006/d2fa287a-cca7-49ce-a771-1bd6e6c876a1)

And we can easily apply access control and authorization:

```python
@app.get("/scalar", response_class=HTMLResponse, include_in_schema=False)
def get_scalar(key: Optional[str] = None) -> str:
    if key == "123456":
        return Scalar(title="Scalar").render()
    else:
        return "Access Denied"
```

Let me know what you think about it and it's designing, thank you 🙏.

**Comment by Kludex** [comment]
Nice 👍

**Reply by hasansezertasan** [reply]
Thank you 😇, just updated the comment and added some screenshots 🚀.

**Reply by mailbaoer** [reply]
Thank you for your excellent work. I noticed some styling issues in the screenshots of the Stoplight doc, so I submitted a PR hoping it can be merged.

**Reply by hasansezertasan** [reply]
> Thank you for your excellent work.

🙏

> I noticed some styling issues in the screenshots of the Stoplight doc, so I submitted a PR hoping it can be merged.

Nice catch, thank you for the PR. I just finished reviewing it and published a new release 🚀!
